feat: parse eyelink events

saeub opened the pull request:
## Description

Parse fixation, saccade, and blink events detected by EyeLink into an `EventDataFrame` and add them to the `GazeDataFrame` returned by `from_asc()`.

Partial fix for #893, alternative implementation to #942.

## Implemented changes

- [x] Parse fixations and saccades
- [x] Add events to `EventDataFrame` (with names `fixation_eyelink` etc.)
- [x] Add additional columns at start of event (e.g. trial information)
- [x] Fix event duration mismatch: https://www.sr-research.com/support/thread-9411.html
- [x] Refactor or remove old blink logic (e.g. data loss ratio, blinks in metadata dictionary)
- [x] Make event parsing optional
- [x] Change default dtype for event onset/offset/duration to `Float64` to support 2kHz ASC files (oversight from #739)
- [ ] Remove `# pylint: disable=too-many-statements` from `parsing.py`

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change is or requires a documentation update

## How Has This Been Tested?

- [x] Add fixations and saccade events to `test_parse_eyelink`
- [x] Test with real ASC file

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

dkrako reviewed (commented):
Great work so far!

I would like to realize #973 before merging this, so we would avoid the breaking change.

I'm indifferent to removing the blink related things from the metadata dictionary.

I would probably leave it there for backwards compatibility and remove it when removing `utils.parsing` as planned in #973

dkrako commented:
>     * [ ]  Add shortcut property to filter fixations, saccades, and blinks (as proposed in [Added Parsing Of BlinksFrom ASC To Data Frame #942 (review)](https://github.com/aeye-lab/pymovements/pull/942#pullrequestreview-2610335217))

I created #968 for this, so you can ignore this issue for this PR.

saeub commented on src/pymovements/utils/parsing.py:
```diff
@@ -213,19 +229,25 @@ def parse_eyelink(
+        'name': [],
+        'onset': [],
+        'offset': [],
+        **{additional_column: [] for additional_column in additional_columns},
```
Yes, I think this is necessary for event properties where gaze and event dataframes need to be joined. I would have lreferred to only include the `trial_columns` in the event dataframe, but `parse_eyelink()` doesn't have that information yet, so I just included all `additional_columns`.

saeub commented:
> I would probably leave it there for backwards compatibility and remove it when removing utils.parsing as planned in https://github.com/aeye-lab/pymovements/issues/973

@dkrako Does this mean that we should keep the blink-related metadata only in the dictionary returned by `utils.parsing.parse_eyelink()`, but remove it from the metadata dictionary returned by `gaze._utils.parsing.parse_eyelink()`?

dkrako commented:
> > I would probably leave it there for backwards compatibility and remove it when removing utils.parsing as planned in #973
> 
> @dkrako Does this mean that we should keep the blink-related metadata only in the dictionary returned by `utils.parsing.parse_eyelink()`, but remove it from the metadata dictionary returned by `gaze._utils.parsing.parse_eyelink()`?

You convinced me in the meeting two weeks ago, that this wasn't really documented and so probably nobody has even used this. I think it's ok to treat this metadata dict as a very experimental feature which is expected to break in the future, as long as we provide an improved functionality (just as you did in this PR).

dkrako commented:
regarding your comment in our minutes:

> the only thing left to decide is how to handle EyeLink's weird way of calculating event durations (https://www.sr-research.com/support/thread-9411-post-36595.html); should we use EyeLink's onset/offset timesteps for the events, or move the onset/offset timestamps to match EyeLink's event durations

I'll just write down some of my thoughts:

Unfortunately we didn't really documented well how we define the offset of an event. We have to improve the documentation there.

Judging from the event detection methods, the offset of an event is defined as the first sample that is not in the event any more. So the offset is exclusive (similar to how python range() works).

If I understand correctly, then EyeLink defines the offset differently: it's the last sample of the event. That's why they increase the duration by one sample-length.

Example event: onset=10, offset=11

in pymovements this would lead to a duration of 1.

In contrast EyeLink assumes the offset to be the last event sample (e.g. inclusive), so the same event would be encoded as onset=10, offset=10
that's why they need to add a complete sample-length to the duration.

This is related to this [comment](https://github.com/aeye-lab/pymovements/issues/688#issuecomment-2116952319) by @behinger 
> I'm not yet sure if you include the events detected by eyelink somewhere, just started checking out the tool. But the automatically detected event-timestamps are rounded down without any way to recover higher precision. This is really a minor thing, as the temporal accuracy to detect such events will never be <1ms - but maybe something to add to a docstring

In conclusion I would say, that we should add a single sample-length to the offset value to counteract the rounding down.

But I'm not fully sure if you'd also expect that, so I'm very open to different points of view @SiQube @prassepaul

saeub commented:
> In conclusion I would say, that we should add a single sample-length to the offset value to counteract the rounding down.

I agree, this is probably the cleanest way to resolve this.

But in that case, I would suggest finishing #887 first, since the sample rate might change within the ASC file, and we need the sampling rate to calculate a sample length.

saeub commented:
@dkrako One more problem: `EventDataFrame`s currently don't support floating-point onsets/offsets for 2kHz ASC files, they are automatically cast (i.e. truncated) to `Int64`. For `GazeDataFrame`s, we already solved this in #739 by changing the default to `Float64`. For `EventDataFrame`s, it doesn't seem as straightforward, judging by the number of failing tests...

I feel like the easiest thing would just be to use `Float64` for all time columns (`time`, `onset`, `offset`, `duration`), but I'm sure there's a reason why you chose integer dtypes in the first place (maybe to prevent floating point precision bugs?). I would appreciate your input here.

saeub commented:
`polars.Duration` sounds neat, I wasn't aware of this!

For now, another option would be to use `Float64` for `onset`, `offset`, and `duration` by default, and convert each of them back to `Int64` if possible, as it's done in `GazeDataFrame`: https://github.com/aeye-lab/pymovements/blob/68c26d01fe14cc0ef561dde19f84e69b795eaa15/src/pymovements/gaze/gaze_dataframe.py#L1532-L1541

dkrako commented:
> For now, another option would be to use `Float64` for `onset`, `offset`, and `duration` by default, and convert each of them back to `Int64` if possible, as it's done in `GazeDataFrame`:

Yes that would be great! It's also ok to work on this in a separate PR.

saeub commented:
@dkrako I added the Float64 → Int64 conversion and the data loss calculation as we discussed.

The data loss ratio behaves slightly differently from before:
- There is no `'unknown'` data loss ratio anymore for edge cases with missing `START`/`END` messages or missing sampling rate. (In my opinion these indicate a corrupt ASC file and would just lead to incorrect data in other places, so as a user I would probably rather get an error message anyway.)
- Changing sampling rates are now supported. The number of expected samples are calculated for each recording block separately.

Unfortunately now pylint is still complaining about too many statements (and too many branches). To be honest, I'm getting quite fed up with these warnings now. We already extracted several code sections into separate functions, but doing this often just makes the code longer and more difficult to navigate instead of increasing readability. What do you think about ignoring these warnings for `parse_eyelink()`?

SiQube commented:
you have my blessing, but @dkrako should decide.

dkrako commented:
> What do you think about ignoring these warnings for `parse_eyelink()`?

I'm perfectly fine with the ignore for now. I think it hints at greater design issues of the `parse_eyelink()` function but that's out of scope of your PR. I will create an issue next week with a proposal on how to generalize `parse_eyelink()` to simplify the code inside and outside the loop by a lot. This will be very useful for #1013.

As far as I skimmed over the code, it looks in good shape already. You can mark it as ready for review after adding the pylint ignores.

saeub commented:
@dkrako
> Judging from the event detection methods, the offset of an event is defined as the first sample that is not in the event any more. So the offset is exclusive (similar to how python range() works).

I just came across some test cases that suggest that the offset is actually inclusive:
https://github.com/aeye-lab/pymovements/blob/9d4bbf348d2f420147b8eee0806696c8b81d8afb/tests/unit/gaze/detect_test.py#L252-L264
(timestamps from 0 to 99, offset is 99)
https://github.com/aeye-lab/pymovements/blob/9d4bbf348d2f420147b8eee0806696c8b81d8afb/tests/unit/gaze/detect_test.py#L355-L373
(timestamps from 1000 to 1099.9, offset is 1099 [rounded down because previously float event offsets were not supported])

dkrako commented:
> @dkrako
> 
> > Judging from the event detection methods, the offset of an event is defined as the first sample that is not in the event any more. So the offset is exclusive (similar to how python range() works).
> 
> I just came across some test cases that suggest that the offset is actually inclusive:

Oh that's a bummer. Of course it's documented nowhere if it's inclusive or not.
I'll need some time to reconsider how to proceed.

saeub commented:
@dkrako I'm wondering whether it's actually useful at all to think about samples as having a "length". If we define a sample as a single point in time, I'd argue we can leave the definition of duration as `offset - onset` and a fixation from time 0 to 99 can have a duration of 99 ms. Then our offsets won't match the ASC files, but our durations will.

saeub commented:
> Both pymovements and EyeLink expects inclusive offsets.

If samples are for a single time point, I would argue that the terms "inclusive" and "exclusive" don't apply anyway. An event with onset `t_1` and offset `t_2` has a duration of `t_2 - t_1`, no matter whether the interval is inclusive `[t_1, t_2]` or exclusive `[t_1, t_2)`.

It seems that EyeLink doesn't see a sample as a single point in time, but as the duration of time in within which the observation was made (at least judging by their [answer in the forum](https://www.sr-research.com/support/thread-9411-post-36595.html)). I.e., in their interpretation, the sample with time stamp `123` lasts from `123.0` to `123.9999999...` at 1000Hz and from `123.0` to `124.99999999...` at 500Hz. I think this is rather counter-intuitive. The current version of the PR changes the offset of the event to be the end-point of the "sample duration", so that we're back to the "point-in-time" definition of samples.

It's a bit confusing, but I hope my explanation makes sense :sweat_smile:

dkrako commented:
Although you're right that samples are only defined for a single time step and don't have any length, the measurement window to acquire a sample does have a length as sampling in the real world is never instantaneous and usually relies on repeated averaged measurements.

What adds confusion is that the eyelink sample timestamp doesn't seem to be a causal definition, as it is forward looking (due to the rounding down). A causal definition would not associate `123` but `124` with the measurement window `123.0 to 123.9999999` (as the sample is calculated after finishing the measurement window, not before). I mean, these are just virtual numbers, but that is strange.

```
|---|---|---|---|---|     sampling intervals
    1   2   3   4   5     causal sample time stamps
1   2   3   4   5         eyelink sample time stamps
```

Here's an example table just to make sure I get the events right (all values as reported from eyelink):

| onset | offset | duration |
| - | - | - |
| 0 | 0 |1|
| 0 | 1 |2 |
| 0 | 10 |11 |

Can you confirm that this is how eyelink reports event offsets and durations?

So they overestimate duration by at most one sampling interval, as there's no way to know that an onset didn't happen at `0.999` and the offset at `2.0` (real duration `1.0000`, eye link duration `2`). The alternative is underestimating duration by at most one sampling interval. That is what we're doing in pymovements and that is what you do when calculating duration as `offset - onset`. So assuming `sample-duration` we use the overestimate method, assuming `point-in-time` we use the underestimate method.

That's why I'm also confused by your comment: you are in favor of a point-in-time definition of samples, but then want to use the `sample-duration` method of duration calculation. This would mean that you favor the EyeLink duration value as being more valid than just the offset value, and that's why you're in favor of using the duration as is and adjusting the offset value, instead of recalculating the duration, right? 

The issue I see is, that this conflates the two duration definitions that are contradicting each other.  Wouldn't it make more sense to stay consistent with the pymovements duration definition? Later on we could improve this by letting users decide how duration is calculated (overestimating by including sampling interval, or underestimating by ignoring sampling interval)?

On the other hand, adjusting the eyelink durations may be more of an issue as this property is much more usable in a direct way than the offset value. Users would probably expect to parse the exact same durations as reported by eyelink and ignore the offset time.

In the end, I think we do not need to solve this issue perfectly for this PR. We can improve this later on in the future. You can decide for yourself which makes more sense more sense to you, as both have their downsides.

What really needs to be done though, is adding a `Notes` section to the `from_asc()` docstring that clarifies how we treat the offset or duration values, and what differs between pymovements and eyelink when we look at the raw data.

dkrako commented:
The first part of my comment is probably irrelevant or maybe even wrong. The remainder of the comment is what's important though (starting with the overestimate/underestimate) and there it is irrelevant if sampling is instantaneous or not.

What we're dealing here is quantization noise for the timestamp and duration values of events.

Note that there isn't any 'correct' way to do this. We won't have enough information to reliably infer the duration of events at the precision of the sampling rate. The precision of a duration will always be halved as it's twice the variance of each sampling timestep (variance being precision). So what we're discussing is just a matter of preference (and consistency). From a statistical point of view, both overestimating and underestimating creates a fixed bias. So alternating the methods or using a random generator would be statistically more optimal than the two simpler methods.

I think it's just fine as you implemented it for now. I'll create a discussion thread on how we want to define the duration of events later on. The best way will probably be to support both preferences.

dkrako commented:
To further clarify as this is important to understand when we're working with event duration on this low level:

Abstract from the eye tracker and the preprocessing and view the complete pipeline just as a single sensor, that outputs the onset of the last fixation of a participant (we want to measure this real world property). We have the real onset of the real fixation `t_r` and the measured fixation onset `t_m`. The sampling rate determines the temporal precision of our sensor: the higher the sampling rate, the smaller the error between `t_r` and `t_m` (meaning the variance of `t_m - t_r` is reduced, leading to higher precision). We call this error noise.

Now let's do the same for a sensor that outputs the duration of the last fixation. We have the real duration of the fixation `l_r` and the measured fixation duration `l_m`. The precision is now twice as low as for our previous sensor: the variance of `l_m - l_r` is twice the variance of `t_m - t_r` as both the onset and the offset timestamps are noisy.

What we do when underestimating or overestimating is assuming a fixed bias to get same precision. But when we picture the probability distributions of `t_m - t_r` we will see that the mean of underestimating will be slightly left and overestimating will lead to slight positive mean value. So we have traded the stochastic error (precision) against a systematic error and thus lost accuracy ("non-biasedness").

But what we actually care about are accurate measurements. Statistically we can usually increase precision by repeated measurements and increasing the population. So as we care most about the accurate measurement of duration we need to sacrifice precision by alternating or randomly choosing over/underestimation. When we picture the distribution of `t_m - t_r`, the mean should be zero, but the variance has doubled.

I took my time to write this down now, as it will probably take some time until I will create a discussion thread or issue for this, as the priority isn't that high (although it is critical) and the topic is complex. I would like to add error distribution plots to fully illustrate my point. Until then I can link to this comment here.

saeub commented:
> What adds confusion is that the eyelink sample timestamp doesn't seem to be a causal definition, as it is forward looking (due to the rounding down).

I'm not sure whether they are looking "forward" or "backward". We really only know that their duration is calculated using this formula: `duration = offset - onset + 1/sampling_rate`.

I agree that it makes sense to have the duration match between between EyeLink and pymovements. But whether we add `1/sampling_rate` to the `offset` or subtract it from `onset` is up to us, I guess.

Another thing that might be even more important is whether the events should be able to overlap or not. EyeLink seems to always start a saccade right after a fixation:

```
393584     58.9   101.8   626.0     0.0 ...
393585     59.2   102.1   626.0     0.0 ...
393586     59.8   102.4   626.0     0.0 ...
EFIX R   392776 393586  811        58.3   100.3     607
SSACC R  393587
393587     60.7   102.5   626.0     0.0 ...
393588     61.6   102.4   626.0     0.0 ...
393589     63.3   101.8   625.0     0.0 ...
```

So if we add `1` to the offset of the fixation or subtract `1` from the onset of the saccade, both of them will have the same timestamp. This means that when calculating event properties like average gaze location, we should only use that sample for one of the events, not both. Currently, event properties are calculated using polar's `is_between` expression with `closed="both"`, meaning that the sample right between a saccade and a fixation will be used for both events:

https://github.com/aeye-lab/pymovements/blob/028bdf608c64f1ddbac25632f34f20c84089eb52/src/pymovements/events/processing.py#L194-L198

Another use case where this could be relevant is when the user wants to check whether a given timestamp is part of a fixation or a saccade. With the current logic, one timestamp could be part of a fixation and a saccade at the same time.

saeub commented:
If you prefer subtracting from the onset instead of adding to the offset, I can change that still (I don't have a strong opinion). Otherwise, I think I'll leave the logic in `parse_eyelink()` as it is now, and add a note about the behavior in the docstring as you suggested.

dkrako commented:
> I agree that it makes sense to have the duration match between between EyeLink and pymovements. But whether we add `1/sampling_rate` to the `offset` or subtract it from `onset` is up to us, I guess.

> Another thing that might be even more important is whether the events should be able to overlap or not. EyeLink seems to always start a saccade right after a fixation:

These two problems are connected and are resolved with the same approach. For clearer language, let's discern duration *value* and duration *function*.

EyeLink and pymovements have two different duration *functions*: EyeLink overestimates durations, pymovements underestimates durations. We are not able to perfectly match both values and semantics of duration between EyeLink and pymovements without either a breaking change or enhancing pymovements with an option to overestimate durations if desired.

That's why although we are able to match duration *values*, we need to mess with the offset value. But the onset and offset values are perfectly correct so what we're doing here is to correct a wrong with an additional wrong because the whole logic is inconsistent. One of the *values* (onset, offset, duration) will always be incorrect when we want to keep the duration *functions* consistent.

So what I would actually prefer most, is keeping the parsed eyelink values completely as is and don't mess with them.
What we are doing here is parsing and changing the data because we choose to calculate duration a different way is unexpected.

This would get rid of the problems that you describe when processing events. The only problem is, that the duration has a different semantic. To make the distinction clear, we could just add a `duration_eyelink` column.

saeub commented:
@dkrako I finally got round to finishing this up. For the blink data loss, I had to keep the "offset+1" logic because the calculations are off otherwise (but the blink events in the EventDataFrame are stored with the original onset and offset values).

dkrako reviewed (approved):
Awesome! Thank you so much for taking care of all the details!
